test/e2e/framework/service/jig.go

```diff
@@ -697,6 +757,108 @@ func (j *TestJig) waitForPodsReady(namespace string, pods []string) error {
 	return nil
 }
 
+func testReachabilityOverServiceName(serviceName string, sp v1.ServicePort, execPod *v1.Pod) {
+	testEndpointReachability(serviceName, sp.Port, sp.Protocol, execPod)
+}
+
+func testReachabilityOverClusterIP(clusterIP string, sp v1.ServicePort, execPod *v1.Pod) {
+	// If .spec.clusterIP is set to "" or "None" for service, ClusterIP is not created, so reachability can not be tested over clusterIP:servicePort
+	isClusterIPV46, err := regexp.MatchString(framework.RegexIPv4+"||"+framework.RegexIPv6, clusterIP)
+	framework.ExpectNoError(err, "Unable to parse ClusterIP: %s", clusterIP)
+	if isClusterIPV46 {
+		testEndpointReachability(clusterIP, sp.Port, sp.Protocol, execPod)
+	}
+}
+func testReachabilityOverNodePorts(nodes *v1.NodeList, sp v1.ServicePort, pod *v1.Pod) {
+	internalAddrs := e2enode.CollectAddresses(nodes, v1.NodeInternalIP)
+	externalAddrs := e2enode.CollectAddresses(nodes, v1.NodeExternalIP)
+	for _, internalAddr := range internalAddrs {
+		testEndpointReachability(internalAddr, sp.NodePort, sp.Protocol, pod)
+	}
+	for _, externalAddr := range externalAddrs {
+		testEndpointReachability(externalAddr, sp.NodePort, sp.Protocol, pod)
+	}
+}
+
+// testEndpointReachability tests reachability to endpoints (i.e. IP, ServiceName) and ports. Test request is initiated from specified execPod.
```

Comment:
Should comment here also that only TCP and UDP are supported here

Reply:
Done